File: src/cat.ts
import { Animal } from "@/animal";

export class Cat extends Animal {
  protected _name: string;
  protected _age: number;

  constructor(parameters: { name: string; age: number }) {
    super();
    this._name = parameters.name;
    this._age = parameters.age;
  }

  public show(): string {
    console.log(`Hi, i'm ${this._name} and i'm ${this._age} yo!`);
    return "a";
  }

  public sound(): void {
    console.log("Meow!");
  }

  public toString(): string {
    return `Cat[name: ${this._name}, age: ${this._age}]`;
  }
}

File: src/cat.rs
use crate::animal::Animal;

pub struct Cat {
    name: String,
    age: u32,
}

impl Animal for Cat {
    fn new(name: String, age: u32) -> Self {
        Cat { name, age }
    }

    fn show(&self) -> () {
        println!("Hi, I'm {}!, I'm {} years old!", self.name, self.age);
    }

    fn sound(&self) -> () {
        println!("Meow!");
    }

    fn to_string(&self) -> String {
        format!("Cat {{ name: {}, age: {} }}", self.name, self.age)
    }
}
